### PR TITLE
Replace Python FSS code by Sycret

### PR DESCRIPTION
**Pull request by tholop**
## Description
This PR replaces the Python implementation for FSS by calls to the Rust library [Sycret](https://github.com/OpenMined/sycret), for faster processing and more compact keys. Work done with @LaRiffle.

In practice, the main changes are:
- DPF and DIF keygen are just wrappers around Sycret, there is no more Python code for FSS
- The mask (alpha) is computed at the Rust level directly instead of outside the keygen
- The format of the keys is more compact: just 2D Numpy arrays of bytes (uint8) instead of list of tuples (easier to pass to Rust through CFFI and faster to parallelize)
- Since we use a keyed PRG instead of a hash function we need one more line to store the AES keys (see Matyas-Meyer-Oseas in Sycret).

Sycret is still in beta but we can start to work on the integration :slightly_smiling_face: 

## Affected Dependencies
- Add Sycret dependency.
- Remove [Shaloop](https://github.com/LaRiffle/shaloop) dependency: we don't need to hash arrays at the Python level anymore, this is done by Sycret!

## How has this been tested?
- Updated the tests in `test_fss.py` to the new primitives format, the logic is unchanged
- Also tested at a lower level in the Sycret library directly
- Tested on Ubuntu 20.04

## Checklist
- [x] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [x] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [x] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [x] My changes are covered by tests

### CODE REVIEW

File: pip-dep/requirements.txt [hunk 1]
```diff
@@ -14,7 +14,7 @@ psutil==5.7.0
 requests~=2.22.0
 RestrictedPython~=5.0
 requests-toolbelt==0.9.1
-shaloop==0.2.1-alpha.11
+sycret==0.0.1-alpha.3
```
Comment (LaRiffle) [inline pip-dep/requirements.txt:17]: Should we use the latests v0.0.1-alpha.6 ?

File: syft/frameworks/torch/mpc/fss.py [hunk 2]
```diff
@@ -81,20 +89,23 @@ def keygen(n_values, op):
...
-            multiprocessing_args.append(process_args)
-
-        with multiprocessing.Pool() as p:
-            partitions = p.starmap(DIF.keygen, multiprocessing_args)
-
-        list_items = _get_items(partitions)
-        return _get_primitives(list_items)
+        return DIF.keygen(n_values=n_values)
+
+        # Multiprocessing in Rust instead.
```
Comment (LaRiffle) [inline syft/frameworks/torch/mpc/fss.py:94]: I guess you can remove all the commented code that you don't use anymore!

File: syft/frameworks/torch/mpc/fss.py [hunk 3]
```diff
@@ -282,6 +322,19 @@ class DPF:
...
+    def eval(b, x, k_b):
+        # x = x.astype(np.uint64)
+        original_shape = x.shape
+        x = x.reshape(-1)
+        flat_result = dpf.eval(b, x, k_b)
+        return flat_result.astype(np.int64).reshape(original_shape)
+        # return result
+
+    @staticmethod
+    def py_keygen(n_values=1):
```
Comment (LaRiffle) [inline syft/frameworks/torch/mpc/fss.py:337]: You can also rm py_keygen and py_eval : they are still accessible in the commit history and the cuda version don't use it :)

File: syft/frameworks/torch/mpc/primitives.py [hunk 4]
```diff
@@ -93,54 +94,24 @@ def get_keys(self, op: str, n_instances: int = 1, remove: bool = True, **kwargs)
...
-
+                keys = primitive_stack[0][0 : n_instances + 1]
+                if remove:
+                    # Keep the AES keys and drop the first n_instance primitives.
+                    remaining_indices = (
+                        np.r_[0, n_instances + 1 : available_instances + 1]
+                        if available_instances > n_instances
+                        else []
+                    )
+                    primitive_stack[0] = primitive_stack[0][remaining_indices]
```
Comment (LaRiffle) [inline syft/frameworks/torch/mpc/primitives.py:114]: Woot woot this is WAY simpler! 🙌

**Review by LaRiffle — CHANGES_REQUESTED** [review]

File: pip-dep/requirements.txt (hunk 1 above)
Comment (tholop) [reply pip-dep/requirements.txt]: Indeed, I moved to `sycret~=0.1.0` so we don't have to change the requirement for every new version -- unless we do an API-breaking change on the Sycret side :)

File: syft/frameworks/torch/mpc/fss.py (hunk 3 above)
Comment (tholop) [reply syft/frameworks/torch/mpc/fss.py]: Yes, I was keeping that for CUDA :) I cleaned this stuff and the comments!

File: syft/frameworks/torch/mpc/primitives.py [hunk 5]
```diff
@@ -214,32 +173,26 @@ def add_primitives(self, types_primitives: dict):
                                 (current_primitives[params][i], primitive)
                             )
             elif op in {"fss_eq", "fss_comp"}:
-                if len(current_primitives) == 0:
-                    setattr(self, op, list(primitives))
+                if len(current_primitives) == 0 or len(current_primitives[0]) == 0:
+                    setattr(self, op, [primitives])
+                    current_primitives = getattr(self, op)
```
Comment (LaRiffle) [inline syft/frameworks/torch/mpc/primitives.py:178]: you can rm this line

File: syft/frameworks/torch/mpc/primitives.py [hunk 6]
```diff
@@ -1,13 +1,14 @@
...
 from typing import List
 
-import numpy as np
 import torch as th
 import syft as sy
 from syft.exceptions import EmptyCryptoPrimitiveStoreError
 from syft.frameworks.torch.mpc.beaver import build_triple
 from syft.workers.abstract import AbstractWorker
 
+import time
```
Comment (LaRiffle) [inline syft/frameworks/torch/mpc/primitives.py:10]: You can rm it

File: syft/frameworks/torch/mpc/primitives.py [hunk 7]
```diff
@@ -93,54 +93,19 @@ def get_keys(self, op: str, n_instances: int = 1, remove: bool = True, **kwargs)
...
-                        )
-
-                        keys.append(prim[get_slice])
-                        if remove:
-                            primitive_stack[i] = prim[remaining_slice]
-
+                keys = primitive_stack[0][0 : n_instances + 1]
+                if remove:
+                    # We throw the whole key array away, not just the keys we used
+                    del primitive_stack[0]
```
Comment (LaRiffle) [inline syft/frameworks/torch/mpc/primitives.py:108]: 🙌

File: syft/frameworks/torch/mpc/primitives.py [hunk 8]
```diff
@@ -181,12 +138,14 @@ def provide_primitives(
 
         builder = self._builders[op]
 
+        t = time.time()
```
Comment (LaRiffle) [inline syft/frameworks/torch/mpc/primitives.py:141]: rm

File: syft/frameworks/torch/mpc/primitives.py [hunk 9]
```diff
@@ -181,12 +138,14 @@ def provide_primitives(
...
         primitives = builder(n_party=len(workers), n_instances=n_instances, **kwargs)
 
         for worker_primitives, worker in zip(primitives, workers):
             worker_types_primitives[worker][op] = worker_primitives
 
         for i, worker in enumerate(workers):
+            t = time.time()
```
Comment (LaRiffle) [inline syft/frameworks/torch/mpc/primitives.py:148]: rm

File: test/torch/mpc/test_fss.py [hunk 10]
```diff
@@ -2,7 +2,9 @@
 import torch as th
 import numpy as np
 
-from syft.frameworks.torch.mpc.fss import DPF, DIF, n
+from syft.frameworks.torch.mpc.fss import DPF, DIF, N
+
+# NOTE: the FSS class is also tested in the Sycret package
```
Comment (LaRiffle) [inline test/torch/mpc/test_fss.py:7]: One test with pre-provided material would be helpful as well!
Suggestion (approx):
```python
sy.local_worker.crypto_store.force_preprocessing = True

sy.local_worker.crypto_store.provide_primitives(
    workers=(data_owner, model_owner),
    n_instances=8,
    op="fss_comp"
)

encryption_kwargs = dict(
    workers=(data_owner, model_owner), # the workers holding shares of the secret-shared encrypted data
    crypto_provider=crypto_provider, # a third party providing some cryptography primitives
    protocol="fss", # the name of the crypto protocol, fss stands for "Function Secret Sharing"
)

t = th.ones(8)
x = t.encrypt(**encryption_kwargs)
z = (x > 0)

sy.local_worker.crypto_store.force_preprocessing = False
```

**Review by LaRiffle — CHANGES_REQUESTED** [review]
Few details + tests :)

File: test/torch/mpc/test_fss.py (hunk 10 above)
Comment (tholop) [reply test/torch/mpc/test_fss.py:7]: Good idea! I tried something here: https://github.com/OpenMined/PySyft/blob/ae0c927f5d94a1920081d4ee377de91fd28322a8/test/torch/mpc/test_fss.py#L143, but I'm not familiar with the high level encrypt()/decrypt() interface so I'm not sure it makes sense.

**Review by LaRiffle — APPROVED** [review]
🙌
